"witch with portals" feature

Kaffeine opened the pull request:
The feature is mutually exclusive to classic "witch as a spawning point" behavior.

Changes:
- Added portals
- Added inf_min_portal_distance to control the min distance between the portals
- Added inf_enable_witch_portals variable that changes the classic witch abilities and lets a witch open portals.
- If portals enabled, then the round will end with `round_end_witch_with_portals` message
- Witch can take (displace) portals.
- A witch without a portal now can spawn zombies (as classic one).
- Only one witch on a map can place a portal at same time.
- An open of a new portal now reopen Exit if both portals existed.
- The Portal Gun has 2 charges. Take of a portal gains 1 charge. Reload time is 5 seconds.
- The Portal owner receives 5 score points on portal used.
- The player who destroy a portal receives 10 score points.

- Implemented portals destruction by weapons:
  - Pistol (all classes)
  - Hammer
  - Shotgun (hero, biologist, medic)
  - Laser Rifle (sniper, hero, scientist, engineer)
  - Explosion (hero rocket launcher, mercenary bombs)
  - Ninja hit (ninja)

The portal takes no damage by the following weapons (on purpose):
- Medic healing laser
- Scientist traps
- Engineer wall

TODO:
- Add and process portals hit points

Possible TODO:
- `/noportal` command to disable teleport use for a player

bretonium commented:
could you please merge master into your branch?

Kaffeine commented:
> could you please merge master into your branch?

Rebased.

yavl commented:
In my opinion, classic witch dot should be default when no portals exist. Most players won't understand what to do with new witch.

Kaffeine commented:
> In my opinion, classic witch dot spawn should be default when no portals exist. Most players won't understand what to do with new witch.

Hm, that is a good possible solution. Thank you for this simple but nice idea!

bretonium commented on src/game/server/gamemodes/mod.cpp:
```diff
@@ -826,6 +826,12 @@ void CGameControllerMOD::DoWincheck()
 
+		if (g_Config.m_InfEnableWitchPortals)
+		{
+			str_format(aBuf, sizeof(aBuf), "round_end_witch_with_portals winner='zombies' survivors='0' duration='%d' round='%d of %d'", Seconds, m_RoundCount+1, g_Config.m_SvRoundsPerMap);
```
please switch the position of this message and of the message about, otherwise it breaks stats

Kaffeine replied:
Hm, yeah, makes sense. 'll do.

bretonium replied:
it seems that we should remove this additional line at all, because it breaks the stats [0]. But it also seems that we can just append something to the `round_end` message. Like `round_end winner='zombies' survivors='0' duration='%d' round='%d of %d' type='new_witch'`

[0] https://github.com/ResamVi/infclass-stats/blob/master/operations/parser.go#L126

Kaffeine replied:
OK

bretonium commented on src/game/server/gamemodes/mod.cpp:
```diff
@@ -916,6 +922,13 @@ void CGameControllerMOD::DoWincheck()
+
+				if (g_Config.m_InfEnableWitchPortals)
+				{
+					str_format(aBuf, sizeof(aBuf), "round_end_witch_with_portals winner='humans' survivors='%d' duration='%d' round='%d of %d'", NumHumans, Seconds, m_RoundCount+1, g_Config.m_SvRoundsPerMap);
```
same here

teoman002 commented:
can you increase the possible radius for placing portals by 20 - 30 %? Its hard to place them without dying near engi laser.

Kaffeine commented:
> can you increase the possible radius for placing portals by 20 - 30 %? Its hard to place them without dying near engi laser.

The distance is the same as for usual witch spawn point. So it is (expected to be) equally hard as it was before.

Kaffeine commented:
Changes:
- Removed static chance for witch spawn, use the classic /witch instead.
- A witch without a portal now can spawn zombies (as classic one).
- Only one witch on a map can place a portal at same time.
- An open of a new portal now reopen Exit if both portals existed.
- Fixed weapon ammo. Now we have exactly two charges.
- Take of a portal now gains 1 charge.
